optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -58,55 +45,283 @@ that uses a functional approach:
+This constraint stream goes over all instances of class `Shift` amongst <<problemFacts,problem facts>> and
+<<planningEntity,planning entities>> in the <<planningProblemAndPlanningSolution,planning problem>>.
+It finds all which are assigned to employee `Ann` and for every such instance (also called "a match"), it adds a soft
+penalty of 1 to the overall <<calculateTheScore,score>>.
+The following figure illustrates this process on a problem with 4 different shifts:
+
+image::ConstraintStreams/constraintStreamIntroduction.png[align="center"]
+
+If any of the instances change during solving, the constraint stream automatically detects it and only recalculates
+the minimum necessary portion of the problem that is affected by the change.
+The following figure illustrates this <<incrementalScoreCalculation,incremental score calculation>>:
+
+image::ConstraintStreams/constraintStreamIncrementalCalculation.png[align="center"]
 
-On the other hand, the ConstraintStreams API uses <<incrementalScoreCalculation,deltas>>
-and supports <<explainingTheScore,justifications>>:
 ////
+TODO Include and describe the following image when PLANNER-1709 is fixed.
+image::ConstraintStreams/constraintStreamJustification.png[align="center"]
+////
+
+
+[[constraintStreamsFirst]]
+== Your First Constraint Stream
+
+To use the ConstraintStreams API in your project, first write a pure Java `ConstraintProvider` implementation like this:
 
 [source,java,options="nowrap"]
 ----
-    private Constraint speakerUnavailableTimeslot(ConstraintFactory factory) {
-        return factory.from(Talk.class)
-                .filter(Talk::hasAnyUnavailableSpeaker)
-                .penalize("Speaker unavailable timeslot",
-                        HardSoftScore.ONE_SOFT,
-                        Talk::getDurationInMinutes);
+    public class MyConstraintProvider implements ConstraintProvider {
+
+        @Override
+        public Constraint[] defineConstraints(ConstraintFactory factory) {
+            return new Constraint[] { // Provide as many constraints as you like. This example only provides one.
+                    firstConstraint(factory)
+            };
+        }
+
+        private Constraint firstConstraint(ConstraintFactory factory) {
+            return factory.from(...) // Choose the initial source of the constraint stream.
+                // Add as many building blocks as you need.
+                .penalize(...); // And finally define the penalty associated with this constraint.
+        }
+
+    }
+----
+
+Then use the following in your solver config:
+
+[source,xml,options="nowrap"]
+----
+    <solver>
+      <scoreDirectorFactory>
+        <constraintProviderClass>com.example.MyConstraintProvider</constraintProviderClass>
+      </scoreDirectorFactory>
+      ...
+    </solver>
+----
+
+
+[[constraintStreamsCardinality]]
+== Constraint Stream Cardinality
+
+Stream cardinality is a measure of how many objects a single constraint match consists of.
```

Comment:
Can we add something about type-safety, which is the motivation why we have multiple cardinality?